main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala

```diff
@@ -0,0 +1,243 @@
+          Lambda(newFormals, replaceVars(body, replaceSyms ++ newMapping), tpe, loc)
+        case e:Hook => e
+        case MkClosure(lambda, freeVars, tpe, loc) => MkClosure(replaceVars(lambda, replaceSyms).asInstanceOf[SimplifiedAst.Expression.Lambda], freeVars, tpe, loc)
+        case e: MkClosureRef => e
+        case ApplyRef(sym, args, tpe, loc) => ApplyRef(sym, args.map(replaceVars(_, replaceSyms)), tpe, loc)
+        case ApplyTail(sym, formals, actuals, tpe, loc) => ApplyTail(sym, formals, actuals.map(replaceVars(_, replaceSyms)), tpe, loc)
+        case ApplyHook(hook, args, tpe, loc) => ApplyHook(hook, args.map(replaceVars(_, replaceSyms)), tpe, loc)
+        case Apply(exp, args, tpe, loc) => Apply(replaceVars(exp, replaceSyms), args.map(a => replaceVars(a, replaceSyms)), tpe, loc)
+        case Unary(op, exp, tpe, loc) => Unary(op, replaceVars(exp, replaceSyms), tpe, loc)
+        case Binary(op, exp1, exp2, tpe, loc) => Binary(op, replaceVars(exp1, replaceSyms), replaceVars(exp2, replaceSyms), tpe, loc)
+        case IfThenElse(exp1, exp2, exp3, tpe, loc) => IfThenElse(replaceVars(exp1, replaceSyms), replaceVars(exp1, replaceSyms), replaceVars(exp3, replaceSyms), tpe, loc)
+        case Let(sym, exp1, exp2, tpe, loc) => Let(replace(sym), replaceVars(exp1, replaceSyms), replaceVars(exp2, replaceSyms), tpe, loc)
+        case LetRec(sym, exp1, exp2, tpe, loc) => LetRec(replace(sym), replaceVars(exp1, replaceSyms), replaceVars(exp2, replaceSyms), tpe, loc)
+        case Is(sym, tag, exp, loc) => Is(sym, tag, replaceVars(exp, replaceSyms), loc)
+        case Tag(sym, tag, exp, tpe, loc) => Tag(sym, tag, replaceVars(exp, replaceSyms), tpe, loc)
+        case Untag(sym, tag, exp, tpe, loc) => Untag(sym, tag, replaceVars(exp, replaceSyms), tpe, loc)
+        case Index(base, offset, tpe, loc) => Index(replaceVars(base, replaceSyms), offset, tpe, loc)
+        case Tuple(elms, tpe, loc) => Tuple(elms.map(e => replaceVars(e, replaceSyms)), tpe, loc)
+        case Existential(fparam, exp, loc) => Existential(fparam, replaceVars(exp, replaceSyms), loc)
+        case Universal(fparam, exp, loc) => Universal(fparam, replaceVars(exp, replaceSyms), loc)
+        case NativeConstructor(constructor, args, tpe, loc) => NativeConstructor(constructor, args.map(a => replaceVars(a, replaceSyms)), tpe, loc)
+        case e:NativeField => e
+        case NativeMethod(method, args, tpe, loc) => NativeMethod(method, args.map(a => replaceVars(a, replaceSyms)), tpe, loc)
+        case e:UserError => e
+        case e:MatchError => e
+        case e:SwitchError => e
+      }
+    }
+
+    /**
+      * Uncurry an expression
+      *
+      * @param tast The expression to uncurry
+      * @return The expression modified to have all curried calls uncurried
+      */
+    def uncurry(tast: SimplifiedAst.Expression, root: SimplifiedAst.Root): SimplifiedAst.Expression = tast match {
```

Comment:
As far as I can tell, this function is not recursive, but applications should be replaced inside subexpressions. E.g. if you have `let f = 42; g(1)(2)(3)` we want to replace g.